### PR TITLE
Fix cluster building for egress gateway

### PR DESCRIPTION
**Pull request by ronenschafferibm**
Egress Gateway fails to forward the requests to the destination.

To demonstrate it, one can do the following:

Configure a `sleep` app inside the mesh to curl to a service outside the mesh (e.g. httpbin.org) going through the egress gateway. (Please see the attached yaml)

The traffic gets to the egress gateway but stops there because the cluster for httpbin.org is configured of type `ORIGINAL_DST`. It should be of type `STRICT_DNS`.
See also: [Original destination](https://www.envoyproxy.io/docs/envoy/v1.6.0/intro/arch_overview/service_discovery#original-destination)

This PR fixes the problem.

I used the following configuration of `Gateway`, `ServiceEntry` and `VirtualService` for the use case:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    # DO NOT CHANGE THESE LABELS
    # The egressgateway is defined in install/kubernetes/helm/istio/values.yaml
    # with these labels
    istio: egressgateway
  servers:
  - port:
      number: 80 #also declared in gateway's deployment files
      name: http
      protocol: HTTP
    hosts:
    - httpbin.org
    #tls:
    #  httpsRedirect: true # sends 302 redirect for http requests
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: httpbin
spec:
  hosts:
    - "httpbin.org"
  ports:
    - number: 80
      name: http-port
      protocol: HTTP
  resolution: NONE
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: route-via-egressgateway
spec:
  hosts:
    - httpbin.org
  gateways:
  # Pinned to both the sidecars (mesh) and egress gateway
  - istio-egressgateway
  - mesh
  http:
    - match:
      - gateways:
        - mesh # from sidecars, route to egress gateway service
      route:
      - destination:
          host: istio-egressgateway.istio-system.svc.cluster.local
        weight: 100
    - match:
      - gateways: # from egress gateway, route to actual external service
        - istio-egressgateway
      route:
      - destination:
          host: httpbin.org
        weight: 100

```

### CODE REVIEW

**Comment by rshriram** [comment]
The issue is with the way you declared the service entry.. It should have resolution: DNS and not resolution NONE

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -85,7 +85,16 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(env model.Environmen
 	for _, service := range services {
 		config := env.DestinationRule(service.Hostname)
 		for _, port := range service.Ports {
-			hosts := buildClusterHosts(env, service, port)
+			var hosts []*core.Address
+			if proxy.Type == model.Router && service.MeshExternal {
+				// In case of Router (e.g. ingress/egress gateway) and external service, we want the service resolution
+				// to be DNS. The router can't use ORIGINAL_DST since, as opposed to a sidecar, it does not get the
+				// traffic by iptables so original_dst is not set.
+				service.Resolution = model.DNSLB
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: see my explanation in the PR. Also, we could have resolution be STATIC and thats perfectly valid in gateway

**Comment by vadimeisenbergibm** [comment]
@rshriram 
> The issue is with the way you declared the service entry.. It should have resolution: DNS and not resolution NONE

How about the following use case: I want to allow traffic to `*.google.com`. I have to use resolution NONE   (so ORIG_DST cluster will be built), haven't I?

So the traffic defined by `*.google.com` cannot be directed thru the Egress Gateway, correct?

Another use case, without the wildcard, suppose I define resolution NONE for my ServiceEntry for `httpbin.org` and everything works fine.

Then later I define a Gateway for httpbin.org, I will have to remember to change the resolution of the ServiceEntry to DNS, correct? So the resolution of the ServiceEntry influences the resolution of both the ServiceEntry and the Gateway, isn't it?

**Comment by rshriram** [comment]
we can't handle *.google.com in egressgateway - today (envoy limitation). And yes, defining resolution as NONE is an error, when using it with gateway. We should probably clarify it in docs..

**Comment by rshriram** [comment]
i.e. we should not auto correct config settings.. it violates user intent (here intent is use the IP I give you).

**Comment by vadimeisenbergibm** [comment]
> we should not auto correct config settings.. it violates user intent (here intent is use the IP I give you).

OK, I see.

**Pull request closed without merging**